File: src/escpos.ts
// Minimal ESC/POS command encoder for thermal receipt printers.
// Pure byte builder — no native deps, cross-compiles into the agent binary.
// Targets Epson TM-T20-compatible printers (the common POS thermal class).

export type EscposChunk = number[] | Uint8Array;

const ESC = 0x1b;
const GS = 0x1d;
const LF = 0x0a;

/** Code page table id → ESC t n. Default 0 (PC437). Add more as needed. */
const CODE_PAGES: Record<number, number> = {
  0: 0, // PC437
  20: 20, // PC860 (Portuguese)
  852: 2, // PC852 (Central European) — varies by printer; left as 2
  1254: 100, // Windows Turkish (printer-dependent)
};

function strBytes(s: string): number[] {
  return Array.from(new TextEncoder().encode(s));
}

export interface PrintLine {
  text: string;
  bold?: boolean;
  align?: 'left' | 'center' | 'right';
  size?: 'normal' | 'double';
}

export interface PrintJob {
  header?: string;
  lines: PrintLine[];
  footer?: string;
  cut?: boolean;
  codePage?: number;
}

export function encodeJob(job: PrintJob): Uint8Array {
  const out: number[] = [];
  const push = (bytes: EscposChunk) => {
    if (bytes instanceof Uint8Array) out.push(...bytes);
    else out.push(...bytes);
  };

  // Init + select code page
  out.push(ESC, 0x40);
  const cp = CODE_PAGES[job.codePage ?? 0] ?? 0;
  out.push(ESC, 0x74, cp);

  const emit = (line: PrintLine) => {
    // align
    const a = line.align === 'center' ? 1 : line.align === 'right' ? 2 : 0;
    out.push(ESC, 0x61, a);
    // bold
    out.push(ESC, 0x45, line.bold ? 1 : 0);
    // size
    if (line.size === 'double') out.push(GS, 0x21, 0x11);
    else out.push(GS, 0x21, 0x00);
    push(strBytes(line.text));
    out.push(LF);
    // reset
    out.push(ESC, 0x61, 0, ESC, 0x45, 0, GS, 0x21, 0x00);
  };

  if (job.header) emit({ text: job.header, bold: true, align: 'center', size: 'double' });
  for (const line of job.lines) emit(line);
  if (job.footer) emit({ text: job.footer, align: 'center' });

  if (job.cut !== false) {
    // Feed 3 lines then partial cut
    out.push(ESC, 0x64, 3, GS, 0x56, 0x42, 0x00);
  }

  return new Uint8Array(out);
}

File: crates/connector-drivers/src/escpos.rs
//! Minimal ESC/POS encoder (ported from TypeScript escpos.ts).

pub struct PrintLine {
    pub text: String,
    pub bold: bool,
    pub align: Align,
    pub size: Size,
}

#[derive(Clone, Copy)]
pub enum Align {
    Left,
    Center,
    Right,
}

#[derive(Clone, Copy)]
pub enum Size {
    Normal,
    Double,
}

pub struct PrintJob {
    pub header: Option<String>,
    pub lines: Vec<PrintLine>,
    pub footer: Option<String>,
    pub cut: bool,
    pub code_page: u8,
}

const ESC: u8 = 0x1b;
const GS: u8 = 0x1d;
const LF: u8 = 0x0a;

pub fn encode_job(job: &PrintJob) -> Vec<u8> {
    let mut out = Vec::new();
    out.extend([ESC, 0x40]);
    out.extend([ESC, 0x74, job.code_page]);

    let emit = |out: &mut Vec<u8>, line: &PrintLine| {
        let align = match line.align {
            Align::Left => 0,
            Align::Center => 1,
            Align::Right => 2,
        };
        out.extend([ESC, 0x61, align]);
        out.extend([ESC, 0x45, if line.bold { 1 } else { 0 }]);
        if matches!(line.size, Size::Double) {
            out.extend([GS, 0x21, 0x11]);
        } else {
            out.extend([GS, 0x21, 0x00]);
        }
        out.extend(line.text.as_bytes());
        out.push(LF);
        out.extend([ESC, 0x61, 0, ESC, 0x45, 0, GS, 0x21, 0x00]);
    };

    if let Some(h) = &job.header {
        emit(
            &mut out,
            &PrintLine {
                text: h.clone(),
                bold: true,
                align: Align::Center,
                size: Size::Double,
            },
        );
    }
    for line in &job.lines {
        emit(&mut out, line);
    }
    if let Some(f) = &job.footer {
        emit(
            &mut out,
            &PrintLine {
                text: f.clone(),
                bold: false,
                align: Align::Center,
                size: Size::Normal,
            },
        );
    }

    if job.cut {
        out.extend([ESC, 0x64, 3, GS, 0x56, 0x42, 0x00]);
    }
    out
}

pub fn encode_drawer_kick() -> Vec<u8> {
    vec![ESC, 0x70, 1, 25, 25]
}
